pootle/apps/accounts/utils.py

```diff
@@ -30,3 +33,245 @@ def get_user_by_email(email):
+            # user.
+            for score_log in submission.scorelog_set.iterator():
+                score_log.delete()
+            submission.save()
+
+    def merge_submitted(self):
+        """Merge submitted_by attribute on units
+        """
+        for unit in self.src_user.submitted.iterator():
+            unit.submitted_by = self.target_user
+            unit.save()
+
+    def merge_suggestions(self):
+        """Merge user attribute on suggestions
+        """
+        # Update user and reviewer on suggestions
+        for suggestion in self.src_user.suggestions.iterator():
+            suggestion.user = self.target_user
+            suggestion.save()
+
+
+class UserPurger(object):
+
+    def __init__(self, user):
+        """Purges user from site reverting any changes that they have made.
+
+        :param user: `User` to purge.
+        """
+        self.user = user
+
+    def purge(self):
+        """Purges user from site reverting any changes that they have made.
+
+        The following steps are taken:
+        - Delete units created by user and without other submissions.
+        - Revert units edited by user.
+        - Revert reviews made by user.
+        - Revert unit comments by user.
+        - Revert unit state changes by user.
+        - Delete any remaining submissions and suggestions.
+        """
+
+        self.remove_units_created()
+        self.revert_units_edited()
+        self.revert_units_reviewed()
+        self.revert_units_commented()
+        self.revert_units_state_changed()
+
+        # Delete remaining submissions.
+        for submission in self.user.submission_set.iterator():
+            submission.delete()
+
+        # Delete remaining suggestions.
+        for suggestion in self.user.suggestions.iterator():
+            suggestion.delete()
+
+    def remove_units_created(self):
+        """Remove units created by user that have not had further
+        activity.
+        """
+
+        # Delete units created by user without submissions by others.
+        for unit in self.user.get_units_created().iterator():
+
+            # Find submissions by other users on this unit.
+            other_subs = unit.submission_set.exclude(submitter=self.user)
+
+            if not other_subs.exists():
+                unit.delete()
+
+    def revert_units_commented(self):
+        """Revert comments made by user on units to previous comment or else
+        just remove the comment.
+        """
+
+        # Revert unit comments where self.user is latest commenter.
+        for unit in self.user.commented.iterator():
+
+            # Find comments by other self.users
+            comments = unit.get_comments().exclude(submitter=self.user)
+
+            if comments.exists():
+                # If there are previous comments by others update the
+                # translator_comment, commented_by, and commented_on
+                last_comment = comments.latest('pk')
```

Comment:
@iafan - in reference to #4025:

here is an example of somewhere that ive had to use ordering by pk - when really i want to order by creation_time

i think its important to be able to order by creation time - because creation_time unlike pk can be manipulated. The fix that i made uses pk only as a secondary ordering when the creation_times match.